sdks/java/core/src/test/java/org/apache/beam/sdk/testing/TestPipelineTest.java

```diff
@@ -52,280 +59,491 @@
+        // @NeedsRunner annotation with a CrashingRunner indicates a configuration issue
+      }
 
-    final PCollection<String> pCollection = pCollection();
-    PAssert.that(pCollection).containsInAnyOrder(DUMMY);
-    pipeline.run().waitUntilFinish();
+      @Category(RunnableOnService.class)
+      @Test
+      public void testConflict_runnableOnServiceAnnotation_expectIllegalStateException()
+          throws Exception {
 
-    // dangling PTransform
-    addTransform(pCollection);
-  }
+        addTransform(pCollection(pipeline));
 
-  @Test
-  public void testNormalFlowWithPAssert() throws Throwable {
-    PAssert.that(pCollection()).containsInAnyOrder(DUMMY);
-    pipeline.run().waitUntilFinish();
-  }
+        // @RunnableOnService annotation with a CrashingRunner indicates a configuration issue
+      }
 
-  @Test
-  public void testAutoAddMissingRunFlow() throws Throwable {
-    PAssert.that(pCollection()).containsInAnyOrder(DUMMY);
-    // missing pipeline#run, but have it auto-added.
-    pipeline.enableAutoRunIfMissing(true);
-  }
+    }
+
+    private static List<Object[]> extractTests(final Class<?> testClass) {
+      final ArrayList<org.junit.runner.Description> testDescriptions = Request.aClass(testClass)
+                                                                              .getRunner()
+                                                                              .getDescription()
+                                                                              .getChildren();
+
+      return FluentIterable.from(testDescriptions)
+                           .transform(new Function<org.junit.runner.Description, Object[]>() {
+
+                             @Override
+                             public Object[] apply(final org.junit.runner.Description description) {
+                               return new Object[] {
+                                   Request.method(testClass, description.getMethodName()),
+                                   description.getTestClass().getSimpleName() + "#"
+                                       + description.getMethodName()
+                               };
+                             }
+                           })
+                           .toList();
+    }
+
+    private TestRule withPossibleExceptions(final String testName) {
+      final String expect = testName.substring(testName.indexOf("_expect"));
+      if (expect.contains(TestPipeline.PipelineRunMissingException.class.getSimpleName())) {
+        final ExpectedException exception = ExpectedException.none();
+        exception.expect(TestPipeline.PipelineRunMissingException.class);
+        return exception;
+      } else if (expect.contains(TestPipeline.AbandonedNodeException.class.getSimpleName())) {
+        final ExpectedException exception = ExpectedException.none();
+        exception.expect(TestPipeline.AbandonedNodeException.class);
+        if (testName.contains(P_TRANSFORM)) {
```

Comment:
@kennknowles thanks for taking a look. I had intentionally jumped through hoops here in order to make these test look exactly like the user would write them, without the stuff WE need to test them (e.g., `thrown.expect` and the likes). This is why I chose to make the expected exceptions external to the tests, users would not expect TestPipeline exceptions to be thrown, instead I used the naming to figure them out outside the test code (i.e. on the meta-layer).

Do you believe we should move away from these tests representing users scenarios 1:1 in favour of having the expected exceptions inside the tests?